sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx

```diff
@@ -0,0 +1,492 @@
+    fOutTree_MCTruth->Branch("run_num",&run_num);
+    fOutTree_MCTruth->Branch("subrun_num",&subrun_num);
+    fOutTree_MCTruth->Branch("UniverseVals", &UniverseVals);
+    fOutTree_MCTruth->Branch("pdg_to_reweight",&fPdg);
+  }
+
+
+  // Read input parameter sets and set up universes
+  size_t n_parsets = FitParSets.size();
+  std::vector<std::string> FitParNames;
+  std::vector<double> FitParNominals;
+  std::vector<double> FitParSigmas;
+  std::map<std::string, double> theNominals;
+
+  fParameterSet.Configure(GetFullName(), mode, fNsims);
+
+  for (size_t i_parset=0; i_parset<n_parsets; ++i_parset){
+    fhicl::ParameterSet theSet = FitParSets.at(i_parset);
+    std::string theName = theSet.get<std::string>("Name");
+    double theNominal = theSet.get<double>("Nominal",1.);
+    double theSigma = theSet.get<double>("Sigma",0.);
+
+    FitParNames.push_back(theName);
+    FitParNominals.push_back(theNominal);
+    FitParSigmas.push_back(theSigma);
+
+    theNominals[theName] = theNominal;
+
+    fParameterSet.AddParameter(theName, theSigma);
+  }
+
+  if (mode=="pm1sigma"){
+    // pm1sigma mode: 0 = +1sigma, 1 = -1sigma of a single parameter. All other parameters at nominal
+    for (size_t i_parset=0; i_parset<n_parsets; ++i_parset){
+      // For each parameter, first create a nominal parameter set (one for +1sigma and one for -1sigma)
+      std::map<std::string, double> tmp_vals_p1sigma(theNominals);
+      std::map<std::string, double> tmp_vals_m1sigma(theNominals);
+      // Now reset the +1sigma and -1sigma values for this parameter set only
+      tmp_vals_p1sigma[FitParNames.at(i_parset)] = FitParNominals.at(i_parset)+FitParSigmas.at(i_parset);
+      tmp_vals_m1sigma[FitParNames.at(i_parset)] = FitParNominals.at(i_parset)-FitParSigmas.at(i_parset);
+
+      if (fDebug){
+        std::cout << "Universe " << i_parset*2 << ": " << FitParNames.at(i_parset) << " = " << FitParNominals.at(i_parset)+FitParSigmas.at(i_parset) << std::endl;
+        std::cout << "Universe " << i_parset*2+1 << ": " << FitParNames.at(i_parset) << " = " << FitParNominals.at(i_parset)-FitParSigmas.at(i_parset) << std::endl;
+      }
+
+      // Finally, add these universes into the vector
+      UniverseVals.push_back(tmp_vals_p1sigma);
+      UniverseVals.push_back(tmp_vals_m1sigma);
+    } // end loop over parsets (i)
+  } // pm1sigma
+  else if (mode=="multisim"){
+    // multisim mode: parameter values sample within the given uncertainty for all parameters simultaneously
+    // Loop over universes j
+    for (unsigned j=0; j<fNsims; j++){
+      // In each multisim universe, loop through all parameters. For each parameter, generate a new random number from Nominal-Sigma to Nominal+Sigma.
+      std::map<std::string, double> tmp_vals;
+      for (size_t i_parset=0; i_parset<n_parsets; ++i_parset){
+        double r = fGaussRandom->fire(0.0,1.0);
+        tmp_vals[FitParNames.at(i_parset)] = FitParNominals.at(i_parset)+(FitParSigmas.at(i_parset)*r);
```

Comment:
What is this supposed to be doing? I guess I just am not sure what `FitParSigmas.at(i_parset)` is supposed to be in this case